### PR TITLE
[Alerting]: Use title instead of slug for retrieving the namespace

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
[POST /api/ruler/{Recipient}/api/v1/rules/{Namespace} ](https://grafana.github.io/alerting-api/#/ruler/RoutePostNameRulesConfig)
used to fail if the namespace contained special characters.
The reason was that we tried to retrieve the namespace from the database using the slug instead of the title.
[That was also problematic](https://github.com/grafana/grafana/pull/32747#discussion_r609530930) because there's a chance multiple namespaces with special characters would result in the same slug.
This fix queries using the folder title which [is unique](https://github.com/grafana/grafana/blob/b0c6cad637022c22a376cc3a3b1bb35a7f0a6064/pkg/services/sqlstore/migrations/dashboard_mig.go#L175) per organisation

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Review by davidmparrott — APPROVED** [review]
LGTM 🎉

**Review by domasx2 — APPROVED** [review]

File: pkg/dashboards/ifaces.go [hunk 1]
```diff
@@ -6,6 +6,8 @@ import "github.com/grafana/grafana/pkg/models"
 type Store interface {
 	// ValidateDashboardBeforeSave validates a dashboard before save.
 	ValidateDashboardBeforeSave(dashboard *models.Dashboard, overwrite bool) (bool, error)
+	// GetFolderByTitle retrieves a dashboard by its title and it's used by unified alerting
```
Comment (aknuds1) [inline pkg/dashboards/ifaces.go:9]: Nit, but I think the following reads better:
```suggestion
	// GetFolderByTitle retrieves a dashboard by its title and is used by unified alerting
```

File: pkg/services/sqlstore/dashboard.go [hunk 2]
```diff
@@ -202,6 +202,32 @@ func (ss *SQLStore) GetDashboard(id, orgID int64, uid, slug string) (*models.Das
 	return &dashboard, nil
 }
 
+// GetDashboardByTile gets a dashboard by its title.
+func (ss *SQLStore) GetFolderByTitle(orgID int64, title string) (*models.Dashboard, error) {
+	if title == "" {
+		return nil, models.ErrDashboardIdentifierNotSet
+	}
+
+	// there is a unique constrain on org_id, folder_id, title
```
Comment (aknuds1) [inline pkg/services/sqlstore/dashboard.go:211]:
```suggestion
	// there is a unique constraint on org_id, folder_id, title
```

File: pkg/services/ngalert/store/alert_rule.go [hunk 3]
```diff
@@ -309,9 +309,9 @@ func (st DBstore) GetRuleGroupAlertRules(query *ngmodels.ListRuleGroupAlertRules
 }
 
 // GetNamespaceUIDBySlug is a handler for retrieving namespace UID by its name.
```
Comment (aknuds1) [inline pkg/services/ngalert/store/alert_rule.go:311]: This needs to be updated.
```suggestion
// GetNamespaceUIDByTitle is a handler for retrieving a namespace UID by its title.
```

File: pkg/services/sqlstore/dashboard.go [hunk 4]
```diff
@@ -202,6 +202,32 @@ func (ss *SQLStore) GetDashboard(id, orgID int64, uid, slug string) (*models.Das
 	return &dashboard, nil
 }
 
+// GetDashboardByTile gets a dashboard by its title.
```
Comment (aknuds1) [inline pkg/services/sqlstore/dashboard.go:205]:
```suggestion
// GetDashboardByTitle gets a dashboard by its title.
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Looks good, but there are some documentation "bugs".

**Review by aknuds1 — APPROVED** [review]

**Pull request merged**